### ISSUE TITLE
ABI incompatibility between geth/web3 and Solidity 0.6.x because of missing payable attribute

### ISSUE THREAD

**Issue by gsalzer** [ethereum/go-ethereum#20760]
#### System information

Geth version: `v1.9.12-unstable-556888c4-20200302/linux-386/go1.13.8`

#### Expected behaviour
Compiling the contract
```
contract Test {
    constructor() public payable {}
}
```
with `solc 0.6.4` and deploying it with some Ether value (using the ABI provided by Solidity) should succeed.

#### Actual behaviour
```
var testContract = web3.eth.contract([{"inputs":[],"stateMutability":"payable","type":"constructor"}]);
var test = testContract.new(
   {
     from: web3.eth.accounts[0], 
     data: '0x6080604052603f8060116000396000f3fe6080604052600080fdfea2646970667358221220a17c728a0be18787dab464f57cd462a5ff641e6e6b01a7d966c4fb2f5c6f53e864736f6c63430006040033', 
     gas: '4700000',
     value: 1
   }, function (e, contract){
    console.log(e, contract);
    if (typeof contract.address !== 'undefined') {
         console.log('Contract mined! address: ' + contract.address + ' transactionHash: ' + contract.transactionHash);
    }
 })
```
results in
```
Error: Cannot send value to non-payable constructor
	at web3.js:2997:23(102)
	at <eval>:1:28(19)
```

#### Analysis

After adding `payable: true` to the ABI in the assignment to variable `testContract`, the deployment works. Solidity versions prior to 0.6.0 had this attribute, and `geth/web3` seems to depend on it to check for payability.

**Comment by gballet** [comment]
This feature would require an upgrade of web3.js. Works has started in this direction with the replace of otto with goja, but promises are required. The future of web3 is currently in flux, and we are considering a switch to ethers.js. We're putting this issue on hold until progress has been made in this multi-week project.

**Comment by garyng2000** [comment]
I simply don't understand this 'removal' for the solc output(serve no purpose, it is not a bug). It breaks many 'old' libraries. I am using Nethereum 2.0(upgrade is always complicated in large projects) and have to re-massage the abi json back to produce the 'constant' and 'payable' keyword.

**Comment by gsalzer** [comment]
Apparently the compiler is still considered experimental, otherwise backwards compatibility would be of more concern.

**Comment by garyng2000** [comment]
may be Microsoft should put that word behind their windows ;-)

**Comment by gsalzer** [comment]
@garyng2000 Maybe you want to add your complaint about the removal of the keywords to the [solc thread](https://github.com/ethereum/solidity/issues/8690).

**Comment by cobordism** [comment]
> After adding `payable: true` to the ABI in the assignment to variable `testContract`, the deployment works. Solidity versions prior to 0.6.0 had this attribute, and `geth/web3` seems to depend on it to check for payability.

thanks for posting the workaround. I had just been stuck on this issue for hours until I finally found your post. Seemingly the issue still exists 2 years later...

Is the geth console no longer the "correct" interface to use?

**Comment by MariusVanDerWijden** [comment]
Yeah the console is very outdated. The issue is, that we can't update to a newer version since the newer web3js uses promises
